File: python/Arrays/Scaler/max_sum_contiguous_subarray_index.py
def maxSubarrayIndex(A):
    n = len(A)

    max_sum = float("-inf")
    curr_sum = float("-inf")

    l = 0
    r = 0

    for i in range(n):
        curr_sum = max(curr_sum+A[i], A[i])
        if A[i] == curr_sum:
            l = i

        if curr_sum > max_sum:
            r = i
        max_sum = max(curr_sum, max_sum)
    
   
    return {
        "max_sum": max_sum,
        "start_index": l,
        "end_index": r,
        "subarray": A[l:r+1]
    }

print(maxSubarrayIndex([-2, 1, -3, 4, -1, 2, 1, -5, 4]))
#TC: O(N)
#SC: O(1)

File: JavaScript/Arrays/Scaler/max_sum_contiguous_subarray_index.js
function maxSubArrayIndex(A) {
  let currSum = Number.NEGATIVE_INFINITY;
  let maxSum = Number.NEGATIVE_INFINITY;

  let l = 0;
  let r = 0;

  for (let i = 0; i < A.length; i++) {
    currSum = Math.max(currSum + A[i], A[i]);
    if (A[i] === currSum) {
      l = i;
    }

    if (currSum > maxSum) {
      r = i;
    }
    maxSum = Math.max(maxSum, currSum);
  }
  return {
    max_sum: maxSum,
    start_index: l,
    end_index: r,
    subarray: A.slice(l, r + 1),
  };
}
console.log(maxSubArrayIndex([-2, 1, -3, 4, -1, 2, 1, -5, 4]));

//TC: O(N)
//SC: O(1)
